packages/kbn-esql-validation-autocomplete/src/shared/resources_helpers.ts

```diff
@@ -10,22 +10,65 @@
 import type { ESQLAst } from '@kbn/esql-ast';
 import type { ESQLCallbacks } from './types';
 import type { ESQLRealField } from '../validation/types';
+import { enrichFieldsWithECSInfo } from '../autocomplete/utils/ecs_metadata_helper';
 
 export function buildQueryUntilPreviousCommand(ast: ESQLAst, queryString: string) {
   const prevCommand = ast[Math.max(ast.length - 2, 0)];
   return prevCommand ? queryString.substring(0, prevCommand.location.max + 1) : queryString;
 }
 
+// ECS Metadata is very static, so info only needs to be fetched once
+export class EcsMetadataCache {
+  private static instance: EcsMetadataCache;
+  private cacheEcsMetadata: Record<string, { type: string; source: string }> | undefined;
+
+  private constructor() {}
+
+  public static getInstance(): EcsMetadataCache {
+    if (!EcsMetadataCache.instance) {
+      EcsMetadataCache.instance = new EcsMetadataCache();
+    }
+    return EcsMetadataCache.instance;
+  }
+
+  public getMetadata(): Record<string, { type: string; source: string }> | undefined {
+    return this.cacheEcsMetadata;
+  }
+
+  public setMetadata(metadata: Record<string, { type: string; source: string }>) {
+    this.cacheEcsMetadata = metadata;
+  }
+}
+
 export function getFieldsByTypeHelper(queryText: string, resourceRetriever?: ESQLCallbacks) {
   const cacheFields = new Map<string, ESQLRealField>();
+
+  const getEcsMetadata = async () => {
+    const cache = EcsMetadataCache.getInstance();
+    if (!cache.getMetadata() && resourceRetriever?.getFieldsMetadata) {
+      // Fetch full list of ECS field
+      const client = await resourceRetriever?.getFieldsMetadata();
```

Comment:
So this cached api returns a smaller payload with only the type of the field. If a field is not in the list it means that doesnt have ECS schema, right @tonyghiani ?

Reply:
Gotcha, thanx Marco!